Feat/user directory free pro functionality

arifulhoque7 opened the pull request:
### PR Title. PRO [PR](https://github.com/weDevsOfficial/wpuf-pro/pull/1336)

Close [1379](https://github.com/weDevsOfficial/wpuf-pro/issues/1379)
**Feature: Add User Directory Module to WP User Frontend Free**

---

### Description

This PR introduces a complete **User Directory** feature to the free version of WP User Frontend, synced with the Pro version's functionality and design patterns.

---

### Summary of Changes

#### 🎯 Core Feature Implementation

**New Module: User Directory** (`modules/user-directory/`)
- Complete directory listing with search, sort, and pagination
- User profile pages with tabbed content (About, Posts, Comments, Files)
- Setup wizard for easy directory configuration
- REST API for AJAX-powered search and pagination
- Pretty URL support for user profiles (`/directory/username`)

#### 📁 Files Added (77 files, +22,194 lines)

**PHP Classes:**
| File | Description |
|------|-------------|
| `modules/user-directory/User_Directory.php` | Main module class and initialization |
| `modules/user-directory/Admin_Menu.php` | Admin menu and settings page |
| `modules/user-directory/Shortcode.php` | Shortcode rendering for directory and profiles |
| Helpers.php | Helper functions for avatars, URLs, tabs, contact info |
| Directory.php | REST API for AJAX search/pagination |
| `modules/user-directory/Post_Type.php` | Custom post type for directory settings |
| `modules/user-directory/PrettyUrls.php` | Rewrite rules for clean profile URLs |
| `modules/user-directory/DirectoryStyles.php` | Dynamic CSS generation |
| `includes/functions/modules.php` | Module helper functions |

**Templates:**
| File | Description |
|------|-------------|
| `views/directory/layout-3.php` | Grid card layout for directory listing |
| row-3.php | Individual user card template |
| `views/directory/template-parts/pagination-shortcode.php` | Pagination component |
| `views/directory/template-parts/search-field.php` | Search input component |
| `views/directory/template-parts/sort-field.php` | Sort dropdown component |
| `views/directory/template-parts/social-icons.php` | Social media icons display |
| layout-2.php | User profile layout with tabs |
| `views/profile/template-parts/about-2.php` | About tab content |
| `views/profile/template-parts/posts-2.php` | Posts tab with pagination |
| `views/profile/template-parts/comments-2.php` | Comments tab with pagination |
| `views/profile/template-parts/file-2.php` | Files tab (Pro feature preview) |
| user-avatar.php | Avatar with initials fallback |

**React Admin UI** (`src/js/user-directory/`):
| File | Description |
|------|-------------|
| `index.js` | Entry point |
| `App.js` | Main application component |
| `components/DirectoryList.js` | Directory listing management |
| `components/DirectoryWizard.js` | Setup wizard component |
| `components/steps/StepBasics.js` | Basic settings step |
| `components/steps/StepLayout.js` | Layout selection step |
| `components/steps/StepProfile.js` | Profile settings step |
| `components/steps/StepTabs.js` | Tab configuration step |
| `components/steps/StepAdvanced.js` | Advanced settings step |
| `components/common/Header.js` | Admin page header |
| `components/common/LayoutCard.js` | Layout preview cards |
| `components/common/MultiSelect.js` | Multi-select dropdown |
| `components/common/Toast.js` | Toast notifications |
| `components/common/DeleteConfirmModal.js` | Delete confirmation modal |

**Build Configuration:**
| File | Description |
|------|-------------|
| `webpack.user-directory.config.js` | Webpack config for admin JS |
| `tailwind.user-directory.config.js` | Tailwind config with wpuf- prefix |
| `postcss.user-directory.config.js` | PostCSS configuration |

**Assets:**
- `assets/css/wpuf-user-directory-free.css` - Admin styles (7,868 lines)
- `assets/css/wpuf-user-directory-frontend.css` - Frontend styles (550 lines)
- `assets/js/wpuf-user-directory-free.js` - Compiled admin bundle
- `assets/js/wpuf-user-directory-frontend.js` - Frontend scripts
- `assets/js/ud-search-shortcode.js` - AJAX search functionality
- `assets/images/user-directory/` - Layout preview images and icons

---

### Key Features

#### 1. Directory Listing
- **Layout**: Grid card style (Layout 3)
- **Search**: Real-time AJAX search by name/email
- **Sort**: Order by name, registration date, ID
- **Pagination**: AJAX-powered with preserved state
- **Avatar**: Custom photo, Gravatar, or initials fallback

#### 2. User Profile
- **Tabs**: About, Posts, Comments, Files
- **Contact Info**: Email, Website, Phone with styled icons
- **Social Links**: Support for common social platforms
- **Bio**: Expandable biography with "Show More/Less"
- **Actions**: Message button (if PM module active), Edit Profile

#### 3. Setup Wizard
- Step-by-step directory creation
- Layout preview cards
- Role filtering
- Tab customization
- Avatar size configuration

#### 4. Profile URL Handling
- Pretty URLs: `/directory/username` or `/directory/123`
- Proper base_url handling for AJAX pagination context
- Shortcode detection for correct URL generation
- Preserved pagination state on "Back to Directory"

---

### Commits

| Hash | Message |
|------|---------|
| `f531687f` | Add free User Directory feature with setup wizard |
| `c1a5ef37` | Update color classes to use ! prefix for importance |
| `547c5309` | Enhance user directory UI and member count logic |
| `8442e2db` | Add file tab to user profile and update build tasks |
| `4f0a05e6` | Add new utility classes for WPUF admin styles |

---

### Testing Checklist

- [ ] Create new directory via setup wizard
- [ ] Directory listing displays users correctly
- [ ] Search filters users in real-time
- [ ] Sort dropdown changes order
- [ ] Pagination works with AJAX
- [ ] "View Profile" link works after pagination
- [ ] Profile page displays all tabs correctly
- [ ] Posts/Comments tabs show user content with pagination
- [ ] Files tab shows Pro upgrade message
- [ ] Avatar displays correctly (image or initials)
- [ ] "Back to Directory" preserves pagination state
- [ ] Pretty URLs work for profiles
- [ ] Mobile responsive design

---

### Screenshots

_Add screenshots of:_
1. Admin setup wizard
2. Directory listing page
3. User profile page
4. Mobile view

---

### Related Issues
- Closes [1218](https://github.com/weDevsOfficial/wpuf-pro/issues/1218)

### Breaking Changes
None. This is a new feature addition.

### Dependencies
- Tailwind CSS (with `wpuf-` prefix)
- WordPress REST API
- React (for admin UI)

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Free User Directory: admin CRUD UI, REST endpoints, shortcode/block listings, pretty profile URLs, responsive directory grids & profile pages with tabs, frontend live search with URL sync, pagination, sorting, and a new React-based admin app with templates and styles.
  * Admin modules split into Free and Pro sections with activation toggles.

* **Bug Fixes**
  * Scoped hover overlay styling and more reliable admin toggle with rollback on error.

* **Chores**
  * Build pipeline updated with user-directory build/watch tasks and npm scripts.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

arifulhoque7 committed:
Add free User Directory feature with setup wizard

Implements a limited User Directory in the free version, including setup wizard UI, REST API, shortcode handler, and directory limit logic. Adds new assets, React components, admin menu, post type registration, and views for directory/profile layouts. Pro-only features are previewed with disabled states and upgrade prompts, following the Pro Preview pattern.

arifulhoque7 committed:
Update color classes to use ! prefix for importance

Added ! prefix to purple color classes in layout helpers and Tailwind config to ensure higher specificity. Updated safelist in Tailwind config to include both prefixed and non-prefixed classes, and adjusted button class usage for consistency.

arifulhoque7 committed:
Enhance user directory UI and member count logic

Improves member count calculation in the REST API and displays live member counts in the directory list. Refines input styles for number fields, updates the toast notification component for better UX, and adds a Pro badge to the 'New Directory' button when the directory limit is reached.

arifulhoque7 committed:
Add file tab to user profile and update build tasks

Introduces a new 'Files' tab to the user profile with grouped file display and filtering, excluding private message attachments. Updates Gruntfile.js to add user directory build/watch tasks and ensures profile size is passed to the template. Also adds the corresponding template file for the new tab.

arifulhoque7 committed:
Add new utility classes for WPUF admin styles

Introduced a wide range of new utility classes to form-builder.css for layout, spacing, sizing, display, border, background, and other style properties. These additions improve flexibility and consistency for admin UI components and support more granular styling options.

arifulhoque7 committed:
Merge branch 'weDevsOfficial:develop' into feat/user-directory-free

arifulhoque7 committed:
Refactor user directory module structure and build config

Moved user directory assets and configs to modules/user-directory, updated Gruntfile.js to reflect new paths, and adjusted build/watch tasks. Renamed Tailwind and Webpack configs, added PostCSS config, and updated package files to support the new module structure.

arifulhoque7 committed:
Refactor advanced step options and update sort values

Updated sort option values from 'ID' to 'id' and adjusted related logic for consistency. Removed free/pro checks from handlers to simplify state updates. Improved UI/UX for avatar and sort options, added Pro-locked features with badges and tooltips, and enhanced descriptions for better clarity.

arifulhoque7 committed:
Update profile layout to use purple theme and improve UI

Changed primary color scheme from emerald to purple across helpers and Tailwind config. Updated profile layout-2 to add a cover photo section, adjust class names, and improve header overlap. Cleaned up the About tab by removing Pro upgrade prompts and related UI elements.

arifulhoque7 committed:
Update primary color scheme to emerald green

Changed Tailwind primary and hover colors from purple to emerald green in tailwind.config.js. Updated related CSS class in App.js to use emerald color for upgrade link, ensuring visual consistency with the new color scheme.

arifulhoque7 committed:
Refactor user avatar logic for consistency and fallback

Updated avatar rendering to use the same logic as the directory listing, prioritizing custom profile photos, then Gravatar, and finally user initials as a fallback. Improved initials calculation and font sizing for better visual consistency. Simplified markup to show either the avatar image or initials, not both, and ensured consistent sizing and styling.

arifulhoque7 committed:
Improve pagination shortcode handling and styling

Add early return if there are no items to display in the pagination shortcode. Also update the current page styling to use layout-specific color classes for better theme consistency.

arifulhoque7 committed:
Refactor user directory module and update styles

Refactored multiple files in the user-directory module, including Admin_Menu.php, Directory.php, Helpers.php, Post_Type.php, PrettyUrls.php, Shortcode.php, and profile layout. Updated Toast.js in the frontend and regenerated several minified CSS files to reflect style changes.

arifulhoque7 committed:
Merge remote-tracking branch 'upstream/develop' into FEAT/user-directory-free

arifulhoque7 committed:
Skip role filter when 'all' is selected in user directory

Updated both Directory API and Shortcode to skip applying the 'role__in' filter if 'all' is present in the roles array. This prevents unnecessary filtering when all roles should be included.

arifulhoque7 commented:
> Directory members are not showing in list (frontend) @arifulhoque7 vai

@Rubaiyat-E-Mohammad vai, now check please. It should be fixed

<img width="1501" height="697" alt="image" src="https://github.com/user-attachments/assets/943cb25c-318a-4f87-bf05-05ac398c6c26" />

Rubaiyat-E-Mohammad commented:
Clicking on view profile is not redirecting to profile page @arifulhoque7 vai

https://github.com/user-attachments/assets/a0352d0b-3a99-44e2-8187-2ff118e56473

arifulhoque7 committed:
Enhance profile tabs and settings compatibility

Updated default and saved settings to include all profile tabs for better Pro compatibility. Improved REST API to merge settings with defaults and existing values, and to sanitize and save all profile tab-related fields. Adjusted frontend JS and wizard defaults to reflect all tabs, and fixed minor issues with sort option selection and profile slug decoding.

arifulhoque7 commented:
> * [x]  Clicking on view profile is not redirecting to profile page @arifulhoque7 vai
> 
>  Screen.Recording.2025-12-26.at.2.30.59.PM.mov 
> * [x]  Searching a nonexistent user and clearing the search box shows an empty directory

done @Rubaiyat-E-Mohammad vai

arifulhoque7 committed:
Merge branch 'develop' into FEAT/user-directory-free-pro-functionality

arifulhoque7 committed:
Merge branch 'develop' into FEAT/user-directory-free-pro-functionality

arifulhoque7 committed:
User Directory: add SingleSelect & UI polish

Introduce a reusable SingleSelect dropdown and refactor User Directory UI/UX. Key changes:

- Add src/js/user-directory/components/common/SingleSelect.js: new reusable single-select dropdown with grouped options and ProBadge support.
- Refactor StepAdvanced to use SingleSelect for sort and gallery size, simplify Pro feature handling, change free avatar default from 192→128 and adjust avatar size isFree flags, and convert several hover-only Pro badges into clickable upgrade links.
- Update StepLayout to show Pro badge on hover with an upgrade link, simplify border/opacity logic and add local hover state.
- Update App.js ProBadge into a clickable upgrade link and reposition/remove some hover-only Pro UI (New Directory button now shows badge inline when limit reached; limit warning block removed/streamlined).
- package.json: ensure user-directory build/dev scripts run npm ci before npm run build/dev.
- Remove get_current_page() from modules/user-directory/Shortcode.php (cleanup).

These changes centralize dropdown logic, improve upgrade CTA consistency, and tidy UI behavior for Pro-locked features.

arifulhoque7 committed:
Use SingleSelect for profile base field

Replace the native <select> for the profile_base field with a SingleSelect component, passing profileBases as options and adapting the onChange to call the existing handleChange ({ target: { name, value } }). Remove the now-unused inputStyle constant and its comment block. This simplifies the markup and delegates select rendering/styling to the SingleSelect component.

Rubaiyat-E-Mohammad commented:
- [ ] Pro badges is showing without hover @arifulhoque7 vai

<img width="809" height="427" alt="image" src="https://github.com/user-attachments/assets/d76a683b-a3df-4fd2-a7fc-7699bfa1e8f8" />

- [ ] Fonts should be bold according to the design feedback. @arifulhoque7 vai

<img width="592" height="876" alt="image" src="https://github.com/user-attachments/assets/6eabc294-4b99-48c8-bddc-2f9a736ba21c" />

arifulhoque7 committed:
Add button/focus classes and update wizard UI

Safelist new button and focus utilities in the user-directory Tailwind config and update DirectoryWizard and StepTabs components. DirectoryWizard: replace verbose button classnames with wpuf-btn-white, add focus ring utilities to Cancel/Prev/Next/Next (submit) buttons for better accessibility, adjust the footer container style to ensure correct left/right positioning and enforce background color. StepTabs: only render the ProBadge on hover to reduce visual clutter. These changes centralize button styles and improve keyboard focus visibility and layout consistency.

arifulhoque7 commented:
done @Rubaiyat-E-Mohammad vai

sapayth commented on assets/js/wpuf-user-directory-frontend.js:
```diff
@@ -0,0 +1,66 @@
+ *
+ * @package WPUF
+ * @subpackage Free/User_Directory
+ * @since 4.3.0
```
do not hardcode version number here, use `WPUF_SINCE`

sapayth commented on modules/user-directory/Api/Directory.php:
```diff
@@ -0,0 +1,750 @@
+ *
+ * @package WPUF
+ * @subpackage Modules/User_Directory
+ * @since 4.3.0
```
do not hardcode version number here or other places, use WPUF_SINCE

sapayth commented on modules/user-directory/views/profile/template-parts/posts-2.php:
```diff
@@ -0,0 +1,226 @@
+    // Add pagination if there are multiple pages
+    if ( $pagination['total_pages'] > 1 ) {
+        // Build base URL and query args for pagination
+        $current_url = ( is_ssl() ? 'https://' : 'http://' ) . $_SERVER['HTTP_HOST'] . $_SERVER['REQUEST_URI'];
```
Direct $_SERVER superglobal access without sanitization. also in:
  - views/directory/layout-3.php:72
  - views/directory/template-parts/pagination-shortcode.php:46
  - views/profile/template-parts/posts-2.php:96
  - views/profile/template-parts/comments-2.php:88

sapayth reviewed (changes requested):
Check the comments/suggestions from coderabbitai. also:
1. SQL Injection via unsanitized orderby in REST API
  modules/user-directory/Api/Directory.php — The orderby parameter from user input is interpolated directly into SQL:
  $orderby_clause = "ORDER BY {$orderby} {$order}";
  The orderby value must be validated against an allowlist of column names, not just sanitized as text.

  2. Direct $_SERVER superglobal access without sanitization (5 files)
  Multiple templates construct URLs from raw $_SERVER['HTTP_HOST'] and $_SERVER['REQUEST_URI'] — these are user-controllable headers:
  - views/directory/layout-3.php:72
  - views/directory/template-parts/pagination-shortcode.php:46
  - views/profile/template-parts/posts-2.php:96
  - views/profile/template-parts/comments-2.php:88

  Fix: Use sanitize_text_field( wp_unslash( $_SERVER['HTTP_HOST'] ) ) or better yet, use WordPress functions like home_url( add_query_arg( null, null ) ).

  3. Missing wp_unslash() on $_GET superglobals
  - views/directory/template-parts/search-field.php:17
  - views/directory/template-parts/row-3.php:135,141,147

  4. Missing output escaping — XSS vectors
  - views/profile/template-parts/posts-2.php:56 — wp_trim_words( get_the_title(), 10 ) echoed without esc_html()
  - views/profile/template-parts/posts-2.php:65 — get_permalink() without esc_url()
  - views/profile/template-parts/comments-2.php:56 — get_comment_link() without esc_url()
  - views/profile/template-parts/comments-2.php:66 — wp_trim_words( $comment->comment_content ) without esc_html()
  - views/profile/layout-2.php:90 — Raw SVG icon output without wp_kses()
  - views/directory/template-parts/row-3.php:39 — $avatar without wp_kses_post()

  5. REST API — Missing nonce verification in frontend JS
  assets/js/ud-search-shortcode.js:57-62 — Calls REST API without X-WP-Nonce header.

  6. REST API permission_callback concerns
  Verify that the directory REST endpoints in Api/Directory.php have proper permission_callback functions — public endpoints should explicitly use __return_true, and any write/admin
  endpoints must check capabilities.

  ---
  HIGH Priority Issues

  7. Hardcoded REST URL in React component
  src/js/user-directory/components/DirectoryList.js:98:
  fetch(`/wp-json/wpuf/v1/user_directory/user_count?...`)
  This ignores config.rest_url and will break on non-standard WordPress installations (subdirectory installs, custom REST prefixes).

  8. ProBadge component defined inside another component body
  src/js/user-directory/components/steps/StepAdvanced.js:87 — Defining a React component inside another component's render function causes it to unmount/remount on every render.
  ProBadge is also duplicated across 6 files — should be a single shared component.

  9. Modal accessibility (WCAG 2.1 Level A violation)
  src/js/user-directory/components/common/DeleteConfirmModal.js:
  - No focus trap
  - No role="dialog" or aria-modal="true"
  - No aria-labelledby
  - "Cancel" and "Delete" button text not wrapped in __() for translation

  10. Pagination uses <span> with onClick — not keyboard accessible
  src/js/user-directory/components/DirectoryList.js:291-305 — Should be <button> elements with aria-label.

  11. useEffect dependency array violation
  src/js/user-directory/App.js:57-59 — fetchDirectories not in deps and not wrapped in useCallback. Will miss updates to config or perPage.

  12. Member count fetching — race conditions & performance
  src/js/user-directory/components/DirectoryList.js:55-115:
  - Multiple setState calls in forEach loop — should be batched
  - No AbortController cleanup for in-flight requests on unmount
  - State updates on unmounted component warnings

  ---
  MEDIUM Priority Issues

  13. in_array() missing strict type comparison
  PHPCS config requires true as 3rd parameter. Found ~11 occurrences across:
  - views/directory/template-parts/social-icons.php:22-25
  - views/profile/layout-2.php:202,233,249,265,283,292
  - views/profile/template-parts/file-2.php:101

  14. Missing ABSPATH check in template files
  Several templates lack if ( ! defined( 'ABSPATH' ) ) { exit; }:
  - layout-3.php, row-3.php, pagination-shortcode.php, posts-2.php, comments-2.php

  15. Functions defined inside template files
  wpuf_ud_build_posts_page_url(), wpuf_ud_build_comments_page_url(), wpuf_ud_build_files_page_url() are defined in template files with function_exists() guards. Move these to
  Helpers.php.

  16. Inline <script> blocks in templates
  views/profile/layout-2.php has 3 inline <script> blocks. Should use wp_add_inline_script() for CSP compliance.

  17. posts_per_page => -1 for user attachments
  views/profile/template-parts/file-2.php:28-29 — Loads ALL attachments then filters in PHP. Performance risk for users with many uploads.

  18. Duplicate pagination markup (DRY violation)
  Nearly identical pagination HTML in 3 template files. Should be a shared partial.

  19. Tooltip ID uses Math.random() on every render
  src/js/user-directory/components/common/Tooltip.js:5 — Breaks aria-describedby association. Use useId() or useRef.

  20. navigator.clipboard.writeText — unhandled Promise rejection
  src/js/user-directory/components/DirectoryList.js:32 — No .catch(), shows "Copied!" optimistically.

  21. No form validation before submission
  src/js/user-directory/components/DirectoryWizard.js:445 — Can submit a directory with empty title.

  22. setTimeout without cleanup in DirectoryWizard
  Lines 411-417, 424-429, 438-441 — Will cause state updates on unmounted component.

  23. mailto: link bug
  views/directory/template-parts/row-3.php:69:
  <a href="mailto:<?php echo esc_url( $user_email ); ?>">
  esc_url() without the mailto: prefix will prepend http://. Should be esc_url( 'mailto:' . $user_email ).

  24. SVG linearGradient ID collision
  views/directory/template-parts/social-icons.php — Instagram SVG gradient ID paint0_linear_5793_2712 will collide when multiple user rows render on the same page.

  25. Tailwind config mixes ESM import with CommonJS module.exports
  modules/user-directory/tailwind.config.js:5-8 — May fail in strict CommonJS environments.

arifulhoque7 committed:
Merge branch 'develop' into FEAT/user-directory-free-pro-functionality

arifulhoque7 committed:
User Directory: file-tabs, sanitization, fetch fixes

Add an external file-tabs script and register/enqueue it for profile files; remove inline tab JS. Improve security and robustness across templates: add ABSPATH checks, use esc_url/esc_html/wp_kses/wp_kses_post where appropriate, use wp_unslash for $_GET input, strict in_array checks, and replace raw $_SERVER URL construction with home_url( add_query_arg( null, null ) ).

Refactor Helpers: change free layout color palette to emerald, update avatar markup to use Tailwind classes, and add utility helpers (pagination URL builders and wpuf_ud_get_image_size_dimensions). Limit attachment queries to avoid memory issues and tighten file grouping checks. Replace duplicated pagination URL helpers in posts/comments templates with shared helpers.

Frontend JS improvements: include X-WP-Nonce header in ud-search-shortcode fetch, add new wpuf-ud-file-tabs.js, and harden React code (useCallback for fetchDirectories, AbortController for member-count fetches, improved batching/loading state for directory member counts, promise-handling fallback for clipboard API). Small doc/version placeholder updates (WPUF_SINCE). These changes aim to improve security, performance, and maintainability.

arifulhoque7 commented:
fixed @sapayth vai

sapayth commented on modules/user-directory/DirectoryStyles.php:
```diff
@@ -0,0 +1,453 @@
+        }
+
+        // Check URL structure
+        $current_uri = trim( $_SERVER['REQUEST_URI'], '/' );
```
use wp_unslash() and sanitization

sapayth commented on modules/user-directory/Api/Directory.php:
```diff
@@ -0,0 +1,741 @@
+        }
+
+        // Flush rewrite rules for pretty URLs
+        flush_rewrite_rules();
```
@arifulhoque7 bhai. `flush_rewrite_rules()` is an expensive function call. Minimize it as much as possible

sapayth commented on modules/user-directory/DirectoryStyles.php:
```diff
@@ -0,0 +1,453 @@
+        ];
+
+        foreach ( $profile_params as $param ) {
+            if ( ! empty( $_GET[$param] ) ) {
```
use `wp_unslash()` and sanitization

sapayth reviewed (changes requested):
- `@since` 4.3.0 hardcoded across all new files. use `WPUF_SINCE` placeholder.
- `flush_rewrite_rules()` is an expensive function call. try to minimize it as much as possible.
- Post_Type.php — show_in_rest => true exposes directory config via WP default REST endpoint `/wp-json/wp/v2/wpuf_user_listing` becomes available. Since this module has its own REST controller, set show_in_rest => false to avoid leaking directory settings to authenticated read-capability users.
- Helpers.php — $layout parameter silently ignored in wpuf_ud_free_get_layout_colors(). The parameter is accepted but never used. Either use it or remove it — the silent ignore will confuse Pro code that passes different layouts expecting different colors.
- modules.php — wpuf_free_deactivate_module() always returns true. Returns true even when the module was never in the active list, masking errors.
- Free_Loader.php — Inline style attributes on new section headings. Move to the module CSS file.

arifulhoque7 committed:
User Directory: sanitize, docs, REST & module fixes

Apply a set of cleanup and safety fixes for the User Directory free module:

- Update @since tags to use WPUF_SINCE across multiple files.
- Add admin CSS for module section titles, pro badge and preview container (replace inline styles with classes).
- Harden input handling by sanitizing usage of $_GET and $_SERVER in DirectoryStyles.
- Remove an unnecessary flush_rewrite_rules() call from REST directory creation and disable REST exposure for the internal wpuf_user_listing post type.
- Improve wpuf_free_deactivate_module: validate module existence, return WP_Error for invalid/not-active cases and return null on success; use wpuf_free_get_modules() for checks.
- Simplify helpers: free layout color API now always returns the emerald palette and related wrapper functions updated accordingly.
- Misc: update docblocks and minor view/template metadata to reflect the changes.

These changes improve security, clarity, and correctness of the free User Directory implementation and replace inline styling with maintainable CSS classes.

arifulhoque7 committed:
User Directory: UI fixes & avatar/tab corrections

Multiple fixes and UI refinements for the User Directory module:

- Ensure per_page is at least 1 to avoid zero-limit pagination.
- Use profile_tabs (instead of default_tabs) in Shortcode and update default profile_tabs order to use 'file' key to match settings.
- Fix Tailwind/PostCSS config paths and include Admin_Menu.php in Tailwind content scan.
- DirectoryWizard: track WP sidebar folded state reactively (MutationObserver), switch to functional setState updates to avoid stale values, and improve fetch error handling.
- MultiSelect: import and use cloneElement instead of React.cloneElement for icon cloning.
- SingleSelect: use semantic <li> for option items.
- StepAdvanced: correct avatar size to 256 and remove pointerEvents style from ProBadge overlay.
- StepTabs: small markup and layout adjustments (heading level change, container maxWidth, minor cleanups).
- avatarSizeHelper: correct mapping for layout-6 to 256.

These changes address UX bugs, ensure consistent config keys, and improve robustness of state updates and error handling.

sapayth reviewed (changes requested):
### Critical Security Issues

| # | Issue | Location | Description |
|---|-------|----------|-------------|
| 1 | **DOM-based XSS via jQuery selector injection** | `views/profile/layout-2.php:390-402` | `urlTab` from URL parameter concatenated into jQuery selector without sanitization. Fix: use `.filter()` + `.data()` comparison, or validate against an allowlist of known tab names. |
| 2 | **Public API exposes user emails without rate limiting** | `Api/Directory.php:96-103` | `search_users` endpoint uses `__return_true` permission callback and `search_columns` includes `user_email`, enabling unauthenticated email harvesting at scale. Fix: remove `user_email` from public search columns, add rate limiting, verify directory post is published. |
| 3 | **`extract()` in template loading — variable injection risk** | `Shortcode.php:357` | `extract( $data )` can overwrite local variables if a filter modifies the `$data` array. Fix: pass `$data` directly to templates and access via `$data['key']`, or at minimum use `extract( $data, EXTR_SKIP )`. |
| 4 | **Unsanitized `profile_tabs` and `excluded_users` stored in DB** | `Api/Directory.php:381-393` | Nested arrays saved verbatim to `post_content` without recursive sanitization. Fix: add dedicated sanitization (recursive `sanitize_text_field` for strings, `absint` for IDs). |
| 5 | **Missing `esc_url()` on pagination links (4 instances)** | `pagination-shortcode.php:96,117,122,134` | `wpuf_ud_build_page_url()` output echoed directly into `href` attributes. Fix: wrap all four instances with `esc_url()`. |
| 6 | **Raw `$_SERVER['HTTP_HOST']` access without sanitization** | `pagination-shortcode.php:50-51` | User-controllable header used for URL building. Fix: use `home_url( add_query_arg( [], false ) )` instead. |
| 7 | **`innerHTML` with server-returned HTML** | `ud-search-shortcode.js:~159,185` | `rows_html` and `pagination_html` injected via `innerHTML`, relying solely on server-side escaping. Fix: document the trust relationship in code comments; long-term consider `DOMPurify`. |

### Standards / Moderate Issues

| # | Issue | Location | Description |
|---|-------|----------|-------------|
| 8 | **Missing `wp_unslash()` on `$_GET` superglobals** | `row-3.php:108,113,118` | `sanitize_text_field()` applied without `wp_unslash()` first. Fix: `sanitize_text_field( wp_unslash( $_GET['orderby'] ) )`. |
| 9 | **`intval()` instead of `absint()` for page number** | `row-3.php:102` | Page number should never be negative. Fix: use `absint( wp_unslash( $_GET['page'] ) )`. |
| 10 | **`sanitize_text_field()` used for URL instead of `esc_url_raw()`** | `Api/Directory.php:301` | `base_url` parameter sanitized as text, not as URL. Fix: use `esc_url_raw()`. |
| 11 | **Search endpoint doesn't verify directory post is published** | `Api/Directory.php:310-314` | Draft or trashed directories can be queried by anyone. Fix: add `'publish' === $post->post_status` check. |
| 12 | **Singleton constructor is `public`** | `User_Directory.php:104` | Defeats singleton pattern. Fix: make constructor `private`. |
| 13 | **Loose comparison where strict is needed** | `DirectoryStyles.php:319` | Uses `==` instead of `===`. Fix: `(int) $post->ID === (int) $id`. |
| 14 | **Hardcoded English strings in JS bypass i18n** | `ud-search-shortcode.js:~167-183` | "No users found" and "Try adjusting" strings not localized. Fix: pass through `wp_localize_script()`. |
| 15 | **`parseInt()` without radix parameter** | `ud-search-shortcode.js:257` | Fix: `parseInt(page, 10)`. |
| 16 | **Global `debounceTimeout` shared across instances** | `ud-search-shortcode.js:13` | Multiple directories on one page share a single timer. Fix: move variable inside `initUserDirectorySearch`. |

### Performance Concerns

| # | Issue | Location | Description |
|---|-------|----------|-------------|
| 17 | **`get_directory_pages()` loads ALL published pages on every `init`** | `PrettyUrls.php:121-126` | `posts_per_page => -1` query with `has_shortcode()` loop on every request. Fix: cache results in a transient, invalidate on `save_post`. |
| 18 | **`flush_rewrite_rules()` on every directory creation and `save_post`** | `Api/Directory.php:237`, `PrettyUrls.php:159` | Expensive `.htaccess` rewrite on every save. Fix: use a deferred flag in `wp_options`, flush once on next `init`. |
| 19 | **`get_directory_count()` loads all post IDs to count** | `User_Directory.php:210-216` | Loads all IDs into memory. Fix: use `wp_count_posts( self::POST_TYPE )`. |
| 20 | **Gravatar HTTP check on first load per user** | `Helpers.php:138-148` | External `wp_remote_head()` per user on first visit. Fix: use background cron to warm cache, or skip Gravatar check entirely. |

### Best Practice Suggestions

| # | Suggestion | Location |
|---|-----------|----------|
| 21 | Use `home_url()` instead of `$_SERVER['HTTP_HOST']` for URL construction | `Helpers.php:220-222` |
| 22 | Add `validate_callback` / `sanitize_callback` in REST route `args` definitions | `Api/Directory.php:53-103` |
| 23 | Move inline `<script>` / `<style>` blocks to enqueued assets | `views/profile/layout-2.php:315-561` |
| 24 | Add ARIA attributes (`role="tab"`, `role="tabpanel"`, `aria-selected`) to tab navigation | `views/profile/layout-2.php:208-213` |
| 25 | Add minimum search length (2+ chars) before firing AJAX | `ud-search-shortcode.js:265-270` |
| 26 | Add `file_exists()` guards on template includes | `layout-3.php:18,27`, `row-3.php:96` |
| 27 | Add PHPCS nonce ignore comments on public-facing `$_GET` accesses | `row-3.php`, `search-field.php` |
| 28 | Add `@since WPUF_SINCE` docblocks to all `DirectoryStyles` methods (15+ missing) | `DirectoryStyles.php` |
| 29 | Validate template path after `apply_filters( 'wpuf_ud_free_template_path' )` to prevent arbitrary file inclusion | `Shortcode.php:370-372` |
| 30 | Use `wp_send_json_error()` with HTTP status codes (e.g., `403`) | `Free_Loader.php` AJAX handler |

arifulhoque7 committed:
Harden user-directory: sanitization & perf

Add extensive sanitization, validation and security hardening across the user-directory module. Key changes:

- Register REST args for directory endpoints with sanitize/validate callbacks and sanitize base_url using esc_url_raw; block user_email from public search columns to reduce email harvesting.
- Recursively sanitize excluded_users input and use sanitized data when extracting IDs.
- Schedule background Gravatar checks via wp_cron (non-blocking) and mark transient as pending to avoid remote requests during page loads.
- Use home_url and esc_url_raw for building profile/base URLs and escape pagination links (esc_url/esc_html/esc_attr).
- Defer flush_rewrite_rules and cache directory pages with transients; invalidate cache on content changes.
- Harden template loading (realpath check) and extract template data using EXTR_SKIP to avoid variable injection.
- Make User_Directory constructor private and optimize get_directory_count using wp_count_posts.
- Sanitize/absint GET parameters in templates and harden JS tab activation with an allowlist and safer selectors to prevent DOM-based XSS.
- Add docblocks/@since annotations and minor refactors to improve robustness and performance.
